### PR TITLE
Adding a PInot/ES response comparator

### PR DESCRIPTION
**Pull request by bowenxia**
**What changed?**
Adding a PInot/ES response comparator

**Why?**
To make sure that the response from pinot and es are the same

**How did you test it?**
Unit test, integration test

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: common/persistence/pinotVisibilityTripleManager.go [hunk 1]
```diff
@@ -274,92 +275,146 @@ func (v *pinotVisibilityTripleManager) ListOpenWorkflowExecutions(
...
+		return comparePinotESCountResponse(ctx, manager, manager2, request, v.logger)
+	}
 	return manager.CountWorkflowExecutions(ctx, request)
 }
 
-func (v *pinotVisibilityTripleManager) chooseVisibilityManagerForRead(domain string) VisibilityManager {
+func (v *pinotVisibilityTripleManager) chooseVisibilityManagerForRead(domain string) (VisibilityManager, VisibilityManager) {
 	var visibilityMgr VisibilityManager
+	var visibilityMgr2 VisibilityManager
+	if v.readModeIsBoth(domain) {
```
[the remark this answers is not in the document]
Comment (bowenxia) [reply common/persistence/pinotVisibilityTripleManager.go:398]: done

File: common/persistence/pinotVisibilityTripleManager.go [hunk 2]
```diff
@@ -274,92 +275,145 @@ func (v *pinotVisibilityTripleManager) ListOpenWorkflowExecutions(
...
+		if v.esVisibilityManager != nil {
+			visibilityMgr = v.esVisibilityManager
+		} else {
+			visibilityMgr = v.pinotVisibilityManager
```
Comment (neil-xie) [inline common/persistence/pinotVisibilityTripleManager.go:402]: We also need to make sure pinotVisibilityManager != nil before we can return

File: common/persistence/pinotResponseComparator.go [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+		panic(fmt.Sprintf("interface to map error in ES/Pinot comparator: %#v", in))
+	}
+
+	return v
+}
+
+func compareSearchAttributes(esSearchAttribute interface{}, pinotSearchAttribute interface{}) error {
+	esAttr, ok := esSearchAttribute.(*types.SearchAttributes)
+	if !ok {
+		panic("interface is not a SearchAttributes! ")
```
Comment (neil-xie) [inline common/persistence/pinotResponseComparator.go:54]: we can cay is not a ES search attribute, same below. So we can know which part has the issue

File: common/persistence/pinotResponseComparator.go [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+)
+
+func interfaceToMap(in interface{}) map[string][]byte {
+	if in == nil || in == "" {
+		return map[string][]byte{}
+	}
+
+	v, ok := in.(map[string][]byte)
+	if !ok {
+		panic(fmt.Sprintf("interface to map error in ES/Pinot comparator: %#v", in))
```
Comment (neil-xie) [inline common/persistence/pinotResponseComparator.go:45]: Actually let's not panic, just return errors. Panicking is not considered a good practice.

File: common/persistence/pinotVisibilityTripleManager.go (hunk 2 above)
Comment (bowenxia) [reply common/persistence/pinotVisibilityTripleManager.go]: done

File: common/persistence/pinotResponseComparator.go (hunk 3 above)
Comment (bowenxia) [reply common/persistence/pinotResponseComparator.go]: done

File: common/persistence/pinotResponseComparator.go (hunk 4 above)
Comment (bowenxia) [reply common/persistence/pinotResponseComparator.go]: done

**Pull request merged**